pkgs/by-name/co/cosmic-greeter/package.nix

```
@@ -48,6 +48,10 @@ rustPlatform.buildRustPackage rec {
     "target/${rust.lib.toRustTargetSpecShort stdenv.hostPlatform}/release/cosmic-greeter"
   ];
 
+  postInstall = ''
+    install -m644 -D cosmic-greeter.toml $out/etc/greetd/cosmic-greeter.toml
```

I think you want to pin all the binaries used in the toml file (AFAICS `cage` and `cosmic-greeter`). Either via search and replace, or since the toml file is small, the whole file via something like `builtins.toTOML`)

---

1. What do you mean by "pin all the binaries"? Add it in `buildInputs` is what I'm assuming but not sure.
2. I can't find any documentation about `builtins.toTOML`.

---

I mean something like `$out/bin/cosmic-greeter` and `${cage}/bin/cage` in the `command` attribute in the `default_session` in the TOML file.

Whoops just checked, `toTOML` was never implemented in nix core (I just skimmed https://github.com/NixOS/nix/issues/3929 and thought since it was closed, that it was implemented...)

In that case (also since `$out` is an env var so not easily injectable in a nix eval context) I'd do something like this:

```suggestion
    sed -i "s|cage -s -- cosmic-greeter|${cage}/bin/cage -s -- $out/bin/cosmic-greeter|" cosmic-greeter.toml
    install -m644 -D cosmic-greeter.toml $out/etc/greetd/cosmic-greeter.toml
```

---

Btw. we may want to make this configurable in NixOS (e.g. [`environment.etc."greetd/cosmic-greeter.toml".text`](https://search.nixos.org/options?channel=23.11&show=environment.etc.%3Cname%3E.text&from=0&size=50&sort=relevance&type=packages&query=environment.etc.)

I'm not sure though...

---

So in that case it should probably rather be in `programs.cosmic-greeter`.

---

Yeah, now that I think of it, it makes sense to have it in `programs.cosmic-greeter`. At which point, I should instead work on getting it in `services.xserver.display-manager.cosmic-greeter`. What do you think?

---

Are we looking at moving it completely then and updating it at the same time?

---

Hmm I think we can include the changes here, can't hurt.

And otherwise, it's probably enough to just use `services.greetd` like this:
```nix
services.greetd = {
  enable = true;
  settings = {
    default_session = {
      command = "${pkgs.cage}/bin/cage -s -- ${pkgs.cosmic-greeter}/bin/cosmic-greeter";
    };
  };
};
```

But adding a whole dedicated service (similar as the `cosmic-greeter.service`) is probably also fine.

---

Only disadvantage with pinning is, that it obviously introduces the additional dependency `cage`.
(In case a different compositor than `cage` is used)

---

Yeah this would be much better done by our NixOS module (WIP at #267099), I don't think there's much value in shipping the upstream greetd toml file (or the upstream systemd service)

> Only disadvantage with pinning is, that it obviously introduces the additional dependency cage.
> (In case a different compositor than cage is used)

I wouldn't worry too much about that, if upstream only supports `cage` then there's no reason to not introduce the dependency

---

I assume so